### PR TITLE
Add advection, hyperdiffusion, and limiters to EDMFX model

### PR DESCRIPTION
**Pull request by dennisYatunin**
## Purpose 

Adds advection, hyperdiffusion, and limiters to the EDFMX model.

## Content

- Modify the `MoistBaroclinicWaveWithEDMF` initial condition so that it sets the initial draft area to 1 and the initial TKE to 0. Add an additional `DryBaroclinicWaveWithEDMF` initial condition, which sets the same values of area and TKE.
- Add `variable_manipulations.jl`, which will contain all useful manipulations of grid-scale and sub-grid-scale variables. Move `divide_by_ρa` to this file, along with the new regularization function `sgs_weight_function`. Move `times_a` to this file, and rename it to `gs_to_sgs`. Add the additional generated functions `specific_gs`, `specific_sgs`, and `matching_subfields`, all of which need to be generated in order to avoid triggering allocations due to type inference failures (ideally, these will not need to be generated in future versions of Julia). Add other helpful functions for computing environment and draft quantities. Ensure that all functions in `variable_manipulations.jl` are thoroughly documented.
- Rename the EDFMX model's `a_min` to `a_half`, in order to reflect how `sgs_weight_function` uses this value.
- Add the specific grid-scale and sub-grid-scale states (`ᶜspecific`, `ᶜspecific⁰`, and `ᶜspecificʲs`) to the precomputed quantities. Use the specific states to greatly simplify the computation of the thermodynamic states. Comment out the modifications to `ᶜK` due to the sub-grid-scale quantities, at least until the modifications are fully detailed in the Overleaf document. Add a hacky roundoff error correction to `ᶜρʲ` in order to ensure that the EDMFX CI examples with a draft area of 1 exhibit the correct behavior (the sub-grid-scale quantities should match the grid-scale quantities when there is no precipitation). Change the diagnostic sub-grid-scale quantities to `ᶜspecific⁺`, `ᶠw⁺`, `ᶜu⁺`, `ᶠu³⁺`, `ᶜK⁺`, `ᶜts⁺`, `ᶜa⁺`, and `ᶜa⁰`. Remove all uses of `getindex` for column fields (e.g., `Y.c.sgsʲs[j]`) in `precomputed_quantities.jl`, since using this syntax involves type piracy through the `TurbulenceConvection` module, and replace them the default `getproperty` syntax instead (e.g., `Y.c.sgsʲs.:($j)`, or `@. Y.c.sgsʲs.:($$j)`).
- Add vertical advection of sub-grid-scale values to `implicit_tendency!`. Rewrite `vertical_transport!` so that it uses the precomputed specific states. Ensure that every quantity in the specific states is vertically transported, with special cases for total energy (since it corresponds to the transport of total enthalpy). Add `edmfx_advection`, and use it for the vertical transport of sub-grid-scale variables. Set its default value to `:none` (central differencing).
- Add advection of sub-grid-scale values to `horizontal_advection_tendency!` and `explicit_vertical_advection_tendency!`. Move the horizontal advection of tracers to `horizontal_tracer_advection_tendency!` in `advection.jl`, and add horizontal advection of sub-grid-scale tracers.
- Add hyperdiffusion of sub-grid-scale values to `hyerdiffusion_tendency!` and `tracer_hyperdiffusion_tendency!`. As with the implicit tendency, rewrite this tendency in terms of the specific states, and ensure that every quantity in the specific states gets hyperdiffusion applied to it. Note that TKE is treated as energy, not as a tracer, so it is affected in `hyerdiffusion_tendency!`.
- Add limiting of sub-grid-scale variables to `limiters_func!`.
- Remove the computation of the nonexistent diagnostic quantities `ᶜq_rai⁰`, `ᶜq_sno⁰`, `ᶜq_rai⁺`, and `ᶜq_sno⁺` for `Microphysics1Moment`.
- Ensure that the modified tendency functions do not trigger allocations due to type inference failures, lowering the allocation limits in `flame.jl`.
- Modify the EDMFX CI examples to use an explicit timestepper (`SSP33ShuOsher`), so that it is not necessary to modify the implicit solver for sub-grid-scale values. Reduce the values of `dt`, `dt_save_to_disk`, and `t_end` to account for the fact that the grid-scale tendencies are no longer treated implicitly. Create a copy of the CI example that runs `MoistBaroclinicWaveWithEDMF` with the EDMFX model, but don't give it a precipitation model. Since precipitation currently only affects grid-scale values, it breaks the symmetry between the grid-scale and sub-grid-scale values. Add another CI example that runs `DryBaroclinicWaveWithEDMF` with the EDFMX model. Ensure that the last two examples have identical evolution of grid-scale and sub-grid-scale values, up to roundoff error.
- Increment the ref counter, since refactoring the tendency functions changes the order of floating point operations, introducing a small MSE relative to the references on the main branch.

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

**Comment by trontrytel** [comment]
@valeriabarra this is what we are trying with advection now

File: src/tendencies/implicit/implicit_tendency.jl [hunk 1]
```diff
@@ -60,91 +49,104 @@ function vertical_transport!(ᶜρcₜ, ᶠu³, ᶜρ, ᶜρc, p, ::Val{:zalesak
...
+            ᶜρaθʲₜ = Yₜc.sgsʲs.:($j).ρaθ
+            ᶠu³ʲ = ᶠu³ʲs.:($j)
+            ᶜρʲ = ᶜρʲs.:($j)
+            ᶜρaθʲ = Yc.sgsʲs.:($j).ρaθ
+            vertical_transport!(ᶜρaθʲₜ, ᶠu³ʲ, ᶜρʲ, ᶜρaθʲ, p, edmfx_upwinding)
+        elseif :ρe_tot in propertynames(Yc)
+            ᶜρae_totʲₜ = Yₜc.sgsʲs.:($j).ρae_tot
+            ᶠu³ʲ = ᶠu³ʲs.:($j)
+            ᶜρʲ = ᶜρʲs.:($j)
+            ᶜρhʲ = ᶜtemp_scalar
```
Comment (szy21) [inline src/tendencies/implicit/implicit_tendency.jl:118]:
```suggestion
            ᶜρahʲ = ᶜtemp_scalar
```

File: src/tendencies/implicit/implicit_tendency.jl [hunk 2]
```diff
@@ -60,91 +49,104 @@ function vertical_transport!(ᶜρcₜ, ᶠu³, ᶜρ, ᶜρc, p, ::Val{:zalesak
...
-    Yₜ.c.uₕ[colidx] .= tuple(zero(eltype(Yₜ.c.uₕ[colidx])))
+    @. Yₜc.uₕ = zero(Yₜc.uₕ)
+
+    @. Yₜf.w =
+        -(ᶠgradᵥ(ᶜp - ᶜp_ref) + ᶠinterp(Yc.ρ - ᶜρ_ref) * ᶠgradᵥ_ᶜΦ) /
+        ᶠinterp(Yc.ρ)
+    for j in 1:n
+        @. Yₜf.sgsʲs[j].w =
+            -(ᶠgradᵥ(ᶜp - ᶜp_ref) + ᶠinterp(ᶜρʲs[j] - ᶜρ_ref) * ᶠgradᵥ_ᶜΦ) /
+            ᶠinterp(Yc.sgsʲs[j].ρa)
```
Comment (szy21) [inline src/tendencies/implicit/implicit_tendency.jl:132]:
```suggestion
            ᶠinterp(Yc.sgsʲs[j].ρ)
```

File: examples/hybrid/cli_options.jl [hunk 3]
```diff
@@ -129,8 +125,11 @@ function parse_commandline()
         default = :none
         "--density_upwinding"
         help = "Denisity upwinding mode [`none` (default), `first_order` , `third_order`, `boris_book`, `zalesak`]"
-        arg_type = Symbol
         default = :none
+        "--edmfx_upwinding"
+        help = "EDMFX upwinding mode [`none` (default), `first_order` , `third_order`, `boris_book`, `zalesak`]"
+        arg_type = Symbol
+        default = :none # TODO: change to :first_order (or higher?)
```
Comment (trontrytel) [inline examples/hybrid/cli_options.jl:132]: When we say `none` does it mean we are using central advection scheme?

File: src/callbacks/callbacks.jl [hunk 4]
```diff
@@ -312,13 +319,41 @@ function save_to_disk_func(integrator)
         )
         if p.atmos.precip_model isa Microphysics1Moment
             ᶜq_rai⁰ =
```
Comment (trontrytel) [inline src/callbacks/callbacks.jl:321]: I would suggest coding only for the 0-moment microphysics. And just throwing an error if someone wants to use EDMFX with 1-moment precipitation. I don't think we have it figured out if/how we want to include rain in the sub-domains.

File: src/tendencies/advection.jl [hunk 5]
```diff
@@ -8,67 +8,105 @@ import ClimaCore.Geometry as Geometry
...
-    (; ᶜu, ᶜK, ᶜΦ, ᶜp, ᶜω³, ᶠω¹², ᶜp_ref) = p
-    point_type = eltype(Fields.local_geometry_field(axes(Y.c)).coordinates)
+    (; ᶜu, ᶜK, ᶜp, ᶜu⁰, ᶜuʲs, ᶜρʲs, ᶜΦ, ᶜp_ref) = p
+    n = n_mass_flux_subdomains(p.atmos.turbconv_model)
 
-    # Mass conservation
-    @. Yₜ.c.ρ -= divₕ(ᶜρ * ᶜu)
+    @. Yₜ.c.ρ -= divₕ(Y.c.ρ * ᶜu)
+    for j in 1:n
+        @. Yₜ.c.sgsʲs.:($$j).ρa -= divₕ(Y.c.sgsʲs.:($$j).ρa * ᶜuʲs.:($$j))
```
Comment (trontrytel) [inline src/tendencies/advection.jl:19]: $$ 😵

File: src/tendencies/implicit/implicit_tendency.jl [hunk 6]
```diff
@@ -60,91 +69,160 @@ function vertical_transport!(ᶜρcₜ, ᶠu³, ᶜρ, ᶜρc, p, ::Val{:zalesak
...
 end
 
-#=
-
-TODO: can we delete this?
-
-# Used for automatically computing the Jacobian ∂Yₜ/∂Y. Currently requires
-# allocation because the cache is stored separately from Y, which means that
-# similar(Y, <:Dual) doesn't allocate an appropriate cache for computing Yₜ.
+implicit_vertical_advection_tendency!(Yₜ, Y, p, t, colidx) =
```
Comment (trontrytel) [inline src/tendencies/implicit/implicit_tendency.jl:72]: Should we still be coding it in a way that allows us to run the EDMFX model fully explicit (for debugging?)

File: src/utils/types.jl [hunk 7]
```diff
@@ -107,9 +100,9 @@ struct EDMFCoriolis{U, V, FT}
 end
 
 struct EDMFX{N, FT}
-    a_min::FT # WARNING: this should never be used outside of divide_by_ρa
+    a_half::FT # WARNING: this should never be used outside of divide_by_ρa
```
Comment (trontrytel) [inline src/utils/types.jl:103]: Should/could we document somewhere what this is doing?

File: examples/hybrid/cli_options.jl (hunk 3 above)
Comment (szy21) [reply examples/hybrid/cli_options.jl:132]: yep

File: src/precomputed_quantities.jl [hunk 8]
```diff
@@ -59,20 +59,104 @@ function precomputed_quantities(atmos, center_space, face_space)
...
+
+Computes the weight of the SGS variables in the linear interpolation used in
+`divide_by_ρa`. This is a continuously differentiable and monotonically
+increasing function of `aʲ` that is equal to 0 when `aʲ ≤ 0`, is equal to 1 when
+`aʲ ≥ 1`, is equal to `1 / 2` when `aʲ = a_half`, grows very rapidly near
+`aʲ = a_half`, and grows very slowly at all other values of `aʲ`.  If `a_half`
+is sufficiently small, this function is essentially equal to 1 for all `aʲ` more
+than a few times larger than `a_half` (up to floating-point precision).
+
+Here is a brief description of how this function is constructed. We need the
```
Comment (trontrytel) [inline src/precomputed_quantities.jl:99]: Would the rest of this long dosctring be more suited for documentation page?

Comment (trontrytel) [reply src/precomputed_quantities.jl:99]: This way we could include a plot of the function, which is probably better than trying to describe what is happening with words

File: src/tendencies/implicit/implicit_tendency.jl [hunk 9]
```diff
@@ -2,149 +2,140 @@
...
+    (; turbconv_model, rayleigh_sponge) = p.atmos
+    (; dt) = p.simulation
+    (; ᶜspecific, ᶠu³, ᶜp, ᶠgradᵥ_ᶜΦ, ᶜρ_ref, ᶜp_ref, ᶜtemp_scalar) = p
+    n = n_mass_flux_subdomains(turbconv_model)
+    if n > 0
+        (; ᶜspecific⁰, ᶜρa⁰, ᶠu³⁰, ᶜspecificʲs, ᶠu³ʲs, ᶜρʲs) = p
+    end
+    ᶜJ = Fields.local_geometry_field(Y.c).J
 
+    ᶜ1 = ᶜtemp_scalar
```
Comment (charleskawczynski) [inline src/tendencies/implicit/implicit_tendency.jl:43]: Please rename `ᶜ1`.

**Comment by dennisYatunin** [comment]
All reviewer comments have been addressed, except for the one about moving the derivation/explanation of the regularization function to the documentation. I'm on vacation through Wednesday, so I'll do that when I get back.

**Pull request merged**